### PR TITLE
Introduce OrderingScheme

### PR DESCRIPTION
**Pull request by sopel39**
Part of: https://github.com/prestodb/presto/pull/9854

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    private final Map<Symbol, SortOrder> orderings;
+
+    @JsonCreator
+    public OrderingScheme(@JsonProperty("orderBy") List<Symbol> orderBy, @JsonProperty("orderings") Map<Symbol, SortOrder> orderings)
+    {
+        requireNonNull(orderBy, "orderBy is null");
+        requireNonNull(orderings, "orderings is null");
+        checkArgument(!orderBy.isEmpty(), "orderBy is empty");
+        checkArgument(orderings.keySet().equals(ImmutableSet.copyOf(orderBy)), "orderBy keys and orderings don't match");
+        this.orderBy = ImmutableList.copyOf(orderBy);
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java]: Can `orderBy` have duplicates?

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java [hunk 2]
```diff
@@ -241,7 +241,7 @@ public Void visitSample(SampleNode node, Void context)
         @Override
         public Void visitSort(SortNode node, Void context)
         {
-            printNode(node, format("Sort[%s]", Joiner.on(", ").join(node.getOrderBy())), NODE_COLORS.get(NodeType.SORT));
+            printNode(node, format("Sort[%s]", Joiner.on(", ").join(node.getOrderingScheme().getOrderBy())), NODE_COLORS.get(NodeType.SORT));
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java]: Consider extracting formatting of `node.getOrderingScheme()` to a separate method. I understand introduction of `OrderingScheme` is all about reuse, so separate method helps reuse.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java [hunk 3]
```diff
@@ -247,9 +247,9 @@ public static PlanMatchPattern sort(PlanMatchPattern source)
         return node(SortNode.class, source);
     }
 
-    public static PlanMatchPattern topN(long count, List<String> orderBy, PlanMatchPattern source)
+    public static PlanMatchPattern topN(long count, ImmutableMap<String, SortOrder> orderingScheme, PlanMatchPattern source)
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: `Map`

also, maybe leave original method, the `topN(long count, List<String> orderBy, PlanMatchPattern source)` signature looks like more convenient

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java [hunk 4]
```diff
@@ -17,32 +17,28 @@
...
 import static com.google.common.base.Preconditions.checkState;
-import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
 
 public class TopNMatcher
         implements Matcher
 {
     private final long count;
-    private final List<PlanTestSymbol> orderBySymbols;
+    private final ImmutableMap<String, SortOrder> orderingScheme;
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java]: `Map`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java [hunk 5]
```diff
@@ -17,32 +17,28 @@
...
 
-    public TopNMatcher(long count, List<PlanTestSymbol> orderBySymbols)
+    public TopNMatcher(long count, ImmutableMap<String, SortOrder> orderingScheme)
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java]: `Map`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java [hunk 6]
```diff
@@ -17,32 +17,28 @@
...
     {
         this.count = count;
-        this.orderBySymbols = ImmutableList.copyOf(orderBySymbols);
+        this.orderingScheme = requireNonNull(orderingScheme, "orderingScheme is null");
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java]: `ImmutableMap.copyOf(`

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java [hunk 7]
```diff
@@ -61,18 +57,12 @@ public MatchResult detailMatches(PlanNode node, StatsProvider stats, Session ses
...
 
-        List<Symbol> expectedOrderBy = orderBySymbols.stream()
-                .map(alias -> alias.toSymbol(symbolAliases))
-                .collect(toImmutableList());
-
-        if (!topNNode.getOrderBy().equals(expectedOrderBy)) {
-            return NO_MATCH;
-        }
-
-        Map<Symbol, SortOrder> expectedOrderings = Maps.toMap(expectedOrderBy, Functions.constant(SortOrder.ASC_NULLS_FIRST));
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java]: I don't understand this change. It looks surprising to me, given the commit is all about replacing two fields in `TopNNode` with single equivalent composite.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java (hunk 1 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java]: I wouldn't assume it cant. Deduplication is `UnaliasSymbolReferences` job.

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java (hunk 2 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java]: TBH I am not sure how `GraphvizPrinter` is used, so I prefer to leave formatting as it was.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java (hunk 3 above)
Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: I think pure `Map` is confusing in this case, because `ImmutableMap` is guaranteed to preserve key insertion order (when iterating). What do you think?

> also, maybe leave original method, the topN(long count, List<String> orderBy, PlanMatchPattern source) signature looks like more convenient

but it implicitly assumes some order. I will leave modified method for now.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java (hunk 4 above)
Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java]: ditto, we care about key insertion order.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java (hunk 7 above)
Comment (sopel39) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java]: As you can see old code implicitly assumed the same ordering for all columns which didn't allow to completely test different sorting cases. I could move this changes to separate commit, but I wonder if it's worth it.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java (hunk 3 above)
Comment (findepi) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: > ImmutableMap is guaranteed to preserve key insertion order (when iterating).

if you're after something that maintains an insertion ordering, use `List`. I know `ImmutableMap` does that too, but making implicit use of that property is an API-obfuscation.

What about:
1. having method that takes 1 symbol and 1 SortOrder
2. having an overload that takes (1 symbol, 1 SortOrder, 2nd symbol, 2nd SortOrder) (if this is still insufficient for test purposes, introduce a builder)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SortNode.java [hunk 8]
```diff
@@ -70,16 +62,10 @@ public PlanNode getSource()
         return source.getOutputSymbols();
     }
 
-    @JsonProperty("orderBy")
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/SortNode.java]: I think you could keep this methods here, just for convenience. Then you would need to any usage of this node.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java (hunk 1 above)
Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/OrderingScheme.java]: AFAIRC there was a bug about that, because ANSI SQL allows that but Presto did not support that. I think @maciejgrzybek fixed that.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java (hunk 3 above)
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: IMO this change is unrelated to the refactoring you are doing, I would revert that.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TopNNode.java [hunk 9]
```diff
@@ -96,16 +89,10 @@ public long getCount()
         return count;
     }
 
-    @JsonProperty("orderBy")
-    public List<Symbol> getOrderBy()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TopNNode.java]: The same as with `SortNode` I would keep these methods.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 10]
```diff
@@ -264,7 +264,9 @@ public Void visitSort(SortNode node, Set<Symbol> boundSymbols)
 
             Set<Symbol> inputs = createInputs(source, boundSymbols);
             checkDependencies(inputs, node.getOutputSymbols(), "Invalid node. Output symbols (%s) not in source plan output (%s)", node.getOutputSymbols(), node.getSource().getOutputSymbols());
-            checkDependencies(inputs, node.getOrderBy(), "Invalid node. Order by dependencies (%s) not in source plan output (%s)", node.getOrderBy(), node.getSource().getOutputSymbols());
+            checkDependencies(inputs, node.getOrderingScheme().getOrderBy(),
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: put each argument in separate new line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 11]
```diff
@@ -859,11 +859,12 @@ private PlanBuilder sort(PlanBuilder subPlan, Optional<OrderBy> orderBy, Optiona
         }
 
         PlanNode planNode;
+        OrderingScheme orderingScheme = new OrderingScheme(orderBySymbols.build(), orderings);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: side note: `orderings` should also be a builder

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java [hunk 12]
```diff
@@ -737,6 +741,27 @@ public String getField()
             return nullOrdering;
         }
 
+        public SortOrder getSortOrder()
+        {
+            checkState(nullOrdering != UNDEFINED, "nullOrdering is undefined");
+            if (ordering == ASCENDING) {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: it looks like it is copied from somewhere... maybe you could extract some utility and reuse it here?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java [hunk 13]
```diff
@@ -61,18 +59,12 @@ public MatchResult detailMatches(PlanNode node, StatsProvider stats, Session ses
...
-
-        if (!topNNode.getOrderingScheme().getOrderings().equals(expectedOrderings)) {
-            return NO_MATCH;
+        for (Ordering ordering : orderBy) {
+            Symbol symbol = Symbol.from(symbolAliases.get(ordering.getField()));
+            OrderingScheme nodeOrderingScheme = topNNode.getOrderingScheme();
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/TopNMatcher.java]: I think you missed the part where you check that order of ordering symbols is correct

**Comment by sopel39** [comment]
applied comments, added ordering schema in windownode.specification

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestSwapAdjacentWindowsBySpecifications.java [hunk 14]
```diff
@@ -96,14 +98,16 @@ public void subsetComesFirst()
                 .on(p ->
                         p.window(new WindowNode.Specification(
                                         ImmutableList.of(p.symbol("a")),
-                                        ImmutableList.of(),
-                                        ImmutableMap.of()),
+                                        new OrderingScheme(
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/TestSwapAdjacentWindowsBySpecifications.java]: OrderingScheme.empty()

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 15]
```diff
@@ -726,10 +726,10 @@ public PhysicalOperation visitTopNRowNumber(TopNRowNumberNode node, LocalExecuti
                     .map(channel -> source.getTypes().get(channel))
                     .collect(toImmutableList());
 
-            List<Symbol> orderBySymbols = node.getOrderBy();
+            List<Symbol> orderBySymbols = node.getOrderingScheme().getOrderBy();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: this belongs to different commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java [hunk 16]
```diff
@@ -772,14 +772,21 @@ public PhysicalOperation visitWindow(WindowNode node, LocalExecutionPlanContext
...
-            List<Symbol> orderBySymbols = node.getOrderBy();
             List<Integer> partitionChannels = ImmutableList.copyOf(getChannelsForSymbols(partitionBySymbols, source.getLayout()));
             List<Integer> preGroupedChannels = ImmutableList.copyOf(getChannelsForSymbols(ImmutableList.copyOf(node.getPrePartitionedInputs()), source.getLayout()));
 
-            List<Integer> sortChannels = getChannelsForSymbols(orderBySymbols, source.getLayout());
-            List<SortOrder> sortOrder = orderBySymbols.stream()
-                    .map(symbol -> node.getOrderings().get(symbol))
-                    .collect(toImmutableList());
+            List<Integer> sortChannels;
+            List<SortOrder> sortOrder;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java]: you can initialize variables here, that way you would not need to have `else` block

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java [hunk 17]
```diff
@@ -737,6 +741,27 @@ public String getField()
             return nullOrdering;
         }
 
+        public SortOrder getSortOrder()
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: Have you seen com.facebook.presto.sql.planner.QueryPlanner#toSortOrder?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 18]
```diff
@@ -777,6 +777,13 @@ private PlanBuilder window(PlanBuilder subPlan, List<FunctionCall> windowFunctio
             List<Symbol> sourceSymbols = subPlan.getRoot().getOutputSymbols();
             ImmutableList.Builder<Symbol> orderBySymbols = ImmutableList.builder();
             orderBySymbols.addAll(orderings.keySet());
+            Optional<OrderingScheme> orderingScheme;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/GatherAndMergeWindows.java [hunk 19]
```diff
@@ -57,12 +58,12 @@ private GatherAndMergeWindows() {}
...
-                                Stream.of(
-                                        new MergeAdjacentWindowsOverProjects(numProjects),
-                                        new SwapAdjacentWindowsBySpecifications(numProjects)))
-                        .collect(toImmutableSet());
+                .boxed()
+                .flatMap(numProjects ->
+                        Stream.of(
+                                new MergeAdjacentWindowsOverProjects(numProjects),
+                                new SwapAdjacentWindowsBySpecifications(numProjects)))
+                .collect(toImmutableSet());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/GatherAndMergeWindows.java]: undo

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 20]
```diff
@@ -338,9 +338,10 @@ public PlanWithProperties visitWindow(WindowNode node, Context context)
             if (!node.getPartitionBy().isEmpty()) {
                 desiredProperties.add(new GroupingProperty<>(node.getPartitionBy()));
             }
-            for (Symbol symbol : node.getOrderBy()) {
-                desiredProperties.add(new SortingProperty<>(symbol, node.getOrderings().get(symbol)));
-            }
+            node.getOrderingScheme().ifPresent(orderingScheme ->
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: 👍

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 21]
```diff
@@ -338,9 +338,10 @@ public PlanWithProperties visitWindow(WindowNode node, Context context)
...
+                    orderingScheme.getOrderBy().stream()
+                            .map(symbol -> new SortingProperty<>(symbol, orderingScheme.getOrdering(symbol)))
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: to me it is weird that you create several properties for single sorting... IMO there should be one property with several symbols, don't you think?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java [hunk 22]
```diff
@@ -209,7 +210,10 @@ public ActualProperties visitWindow(WindowNode node, List<ActualProperties> inpu
             ActualProperties properties = Iterables.getOnlyElement(inputProperties);
 
             // If the input is completely pre-partitioned and sorted, then the original input properties will be respected
-            if (ImmutableSet.copyOf(node.getPartitionBy()).equals(node.getPrePartitionedInputs()) && node.getPreSortedOrderPrefix() == node.getOrderBy().size()) {
+            Optional<OrderingScheme> orderingScheme = node.getOrderingScheme();
+            if (ImmutableSet.copyOf(node.getPartitionBy()).equals(node.getPrePartitionedInputs())
+                    && orderingScheme.isPresent()
+                    && node.getPreSortedOrderPrefix() == orderingScheme.get().getOrderBy().size()) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: the code you are changing looks looks like a hack

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java [hunk 23]
```diff
@@ -231,9 +235,11 @@ public ActualProperties visitWindow(WindowNode node, List<ActualProperties> inpu
...
                 localProperties.add(new GroupingProperty<>(node.getPartitionBy()));
             }
-            for (Symbol column : node.getOrderBy()) {
-                localProperties.add(new SortingProperty<>(column, node.getOrderings().get(column)));
-            }
+
+            orderingScheme.ifPresent(scheme ->
+                    scheme.getOrderBy().stream()
+                            .map(column -> new SortingProperty<>(column, scheme.getOrdering(column)))
+                            .forEach(localProperties::add));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java [hunk 24]
```diff
@@ -286,8 +292,8 @@ public ActualProperties visitTopNRowNumber(TopNRowNumberNode node, List<ActualPr
 
             ImmutableList.Builder<LocalProperty<Symbol>> localProperties = ImmutableList.builder();
             localProperties.add(new GroupingProperty<>(node.getPartitionBy()));
-            for (Symbol column : node.getOrderBy()) {
-                localProperties.add(new SortingProperty<>(column, node.getOrderings().get(column)));
+            for (Symbol column : node.getOrderingScheme().getOrderBy()) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: wrong commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java [hunk 25]
```diff
@@ -571,7 +574,7 @@ public PlanNode visitTopNRowNumber(TopNRowNumberNode node, RewriteContext<Set<Sy
             ImmutableSet.Builder<Symbol> expectedInputs = ImmutableSet.<Symbol>builder()
                     .addAll(context.get())
                     .addAll(node.getPartitionBy())
-                    .addAll(node.getOrderBy());
+                    .addAll(node.getOrderingScheme().getOrderBy());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java]: wrong commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TopNRowNumberNode.java [hunk 26]
```diff
@@ -56,6 +55,7 @@ public TopNRowNumberNode(
 
         requireNonNull(source, "source is null");
         requireNonNull(specification, "specification is null");
+        checkArgument(specification.getOrderingScheme().isPresent(), "specification orderingScheme is absent");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TopNRowNumberNode.java]: wrong commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java [hunk 27]
```diff
@@ -774,7 +774,7 @@ public Void visitTopNRowNumber(TopNRowNumberNode node, Integer indent)
         {
             List<String> partitionBy = Lists.transform(node.getPartitionBy(), Functions.toStringFunction());
 
-            List<String> orderBy = Lists.transform(node.getOrderBy(), input -> input + " " + node.getOrderings().get(input));
+            List<String> orderBy = Lists.transform(node.getOrderingScheme().getOrderBy(), input -> input + " " + node.getOrderingScheme().getOrderings().get(input));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/planPrinter/PlanPrinter.java]: use `Stream`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 28]
```diff
@@ -185,7 +191,11 @@ public Void visitTopNRowNumber(TopNRowNumberNode node, Set<Symbol> boundSymbols)
 
             Set<Symbol> inputs = createInputs(source, boundSymbols);
             checkDependencies(inputs, node.getPartitionBy(), "Invalid node. Partition by symbols (%s) not in source plan output (%s)", node.getPartitionBy(), node.getSource().getOutputSymbols());
-            checkDependencies(inputs, node.getOrderBy(), "Invalid node. Order by symbols (%s) not in source plan output (%s)", node.getOrderBy(), node.getSource().getOutputSymbols());
+            checkDependencies(
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: wrong commit

File: presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java [hunk 29]
```diff
@@ -274,7 +279,9 @@ public Void visitTopNRowNumber(TopNRowNumberNode node, Void context)
         {
             printNode(node,
                     "TopNRowNumber",
-                    format("partition by = %s|order by = %s|n = %s", Joiner.on(", ").join(node.getPartitionBy()), Joiner.on(", ").join(node.getOrderBy()), node.getMaxRowCountPerPartition()),
+                    format("partition by = %s|order by = %s|n = %s",
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/util/GraphvizPrinter.java]: wrong commit

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SpecificationProvider.java [hunk 30]
```diff
@@ -46,19 +48,28 @@
...
+                            .stream()
+                            .map(alias -> alias.toSymbol(aliases))
+                            .collect(toImmutableList()),
+                    orderings
+                            .entrySet()
+                            .stream()
+                            .collect(toImmutableMap(entry -> entry.getKey().toSymbol(aliases), Map.Entry::getValue))));
+        }
+        else {
+            orderingScheme = Optional.empty();
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SpecificationProvider.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java (hunk 22 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: TBH it probably requires to understand window operator more deeply. For now I will preserver previous semantics:
```
node.getPreSortedOrderPrefix() == node.getOrderBy().size())
```
which translates to:
```
node.getPreSortedOrderPrefix() == 0 || node.getPreSortedOrderPrefix() == orderingScheme.get().getOrderBy().size()
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java (hunk 24 above)
Comment (sopel39) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PropertyDerivations.java]: ditto, `WindowNode.Specification` changes propagate to other nodes.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java [hunk 31]
```diff
@@ -474,8 +474,8 @@ public PlanWithProperties visitSort(SortNode node, Context context)
                 // skip the SortNode if the local properties guarantee ordering on Sort keys
                 // TODO: This should be extracted as a separate optimizer once the planner is able to reason about the ordering of each operator
                 List<LocalProperty<Symbol>> desiredProperties = new ArrayList<>();
-                for (Symbol symbol : node.getOrderBy()) {
-                    desiredProperties.add(new SortingProperty<>(symbol, node.getOrderings().get(symbol)));
+                for (Symbol symbol : node.getOrderingScheme().getOrderBy()) {
+                    desiredProperties.add(new SortingProperty<>(symbol, node.getOrderingScheme().getOrdering(symbol)));
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java]: either commits are not sorted or this commit does not compile as there is no `OrderingScheme#getOrdering` method yet
